design/baremetal-operator/node-specific-kernel-args.md

```diff
@@ -0,0 +1,166 @@
+array as a value.
+
+The proposed BMH looks like the following:
+
+```yaml
+apiVersion: metal3.io/v1alpha1
+kind: BareMetalHost
+metadata:
+ name: bm0
+spec:
+ online: true
+ bmc:
+   address: idrac://192.168.122.1:6230/
+   credentialsName: bm0-bmc-secret
+ preProvisionKernelArgs:
+   - --timeout
+   - 60000
+   - systemd.journald.forward_to_console=yes
+   - ipa-debug=1
+   ...
+   ...
+   ...
+ bootMACAddress: 52:54:00:b7:b2:6f
+...
+...
+...
+
+```
+
+The example contains a snippet of a common IPA kernel argument list.
+The `preProvisionKernelArgs` list would be received by the `Provisioner`
+interface as is. The handling of the list would be up to the particular
+implementations of the `Provisioner` interface. In case of the `Ironic`
+provisioner the elements from the `preProvisionKernelArgs` would be
+concatenated into a single string such that the string would start with the
+first element (lowest index) from the list and would end with the last
+(highest index) element. The elements between the first and last element would
+be concatenated in an ascending order based on their index number in the array.
+
+Taken the BMH snippet from above the kernel argument list generated by
+the `Ironic provisioner` and posted to the Ironic Node API  would look like
+`--timeout 60000 systemd.journald.forward_to_console=yes ipa-debug=1 ...`.
+
+### Risks and Mitigations
+
+Providing a faulty list kernel arguments might result in boot issues
+or unexpected behavior during the life cycle of the pre-provisioning image.
+
+The user has to be aware that the kernel arguments supplied here will fully or
+partially overwrite the default arguments depending on how the particular
+`Provisioner` is handling the argument list. In case of Ironic provisioner
+the `preProvisionKernelArgs` will completely replace the default argument
+list used by Ironic.
+
+## Design Details
+
+Most of the design is very straightforward an easy to combine with the
+reconcile logic, the support for node specific pre-provisioning kernel args
+also already exist in Ironic, so right after the feature is implemented in BMO
+it can be used.
+
+The only challenge is the process of updating/cleaning the previously applied
+kernel argument list, this cleanup/update process highly depends on the
+particular provisioner but we might be able to leverage the
+`kubectl.kubernetes.io/last-applied-configuration` annotation or some other
+method to avoid initiating redundant kernel argument update,sync or cleaning
+operations.
+
+### Work Items
+
+- Implement a string array type optional spec field named
+  `preProvisionKernelArgs`
+- Extend the generic `Provisioner` interface to support the new optional field
+- Implement support for the `preProvisionKernelArgs` in the Ironic driver
+- Implement unit tests
+- Add a check to the existing BMO e2e workflow that logs in to an active IPA
+  via ssh or serial console and checks whether the custom kernel args were
+  propagated properly or not.
```

Comment:
:+1: 
I have recently learned also that it is possible to use the kernel command line to [configure the network](https://www.kernel.org/doc/Documentation/filesystems/nfs/nfsroot.txt). Maybe we could use that to verify also, by configuring a static IP outside of DHCP range? If IPA becomes reachable at that IP we can consider the test passed.